pkg/mads/server/server.go

```diff
@@ -1,33 +1,200 @@
+		return
+	}
+	m.passthrough.ServeHTTP(writer, req)
+}
+
+type HttpService interface {
+	RegisterRoutes(ws *restful.WebService)
+}
+
+type GrpcService interface {
+	RegisterWithGrpcServer(server *grpc.Server)
+}
+
+var (
+	_ component.Component = &muxServer{}
+)
+
+func (s *muxServer) createGRPCServer() *grpc.Server {
+	grpcOptions := []grpc.ServerOption{
+		grpc.MaxConcurrentStreams(grpcMaxConcurrentStreams),
+		grpc.KeepaliveParams(keepalive.ServerParameters{
+			Time:    grpcKeepAliveTime,
+			Timeout: grpcKeepAliveTime,
+		}),
+		grpc.KeepaliveEnforcementPolicy(keepalive.EnforcementPolicy{
+			MinTime:             grpcKeepAliveTime,
+			PermitWithoutStream: true,
+		}),
+	}
+	grpcOptions = append(grpcOptions, s.metrics.GRPCServerInterceptors()...)
+	grpcOptions = append(grpcOptions, grpc.MaxConcurrentStreams(grpcMaxConcurrentStreams))
+	server := grpc.NewServer(grpcOptions...)
+
+	for _, service := range s.grpcServices {
+		service.RegisterWithGrpcServer(server)
+	}
+
+	s.metrics.RegisterGRPC(server)
+
+	return server
+}
+
+func (s *muxServer) createHttpServicesHandler() http.Handler {
+	container := restful.NewContainer()
+	promMiddleware := middleware.New(middleware.Config{
+		Recorder: http_prometheus.NewRecorder(http_prometheus.Config{
+			Registry: s.metrics,
+			Prefix:   "mads_server",
+		}),
+	})
+	promFilterFunc := util_prometheus.MetricsHandler("", promMiddleware)
+
+	for _, service := range s.httpServices {
+		ws := new(restful.WebService)
+		ws.Filter(promFilterFunc)
+		service.RegisterRoutes(ws)
+		container.Add(ws)
+	}
+
+	return container
+}
+
+func (s *muxServer) Start(stop <-chan struct{}) error {
+	server := &http.Server{
+		Addr: fmt.Sprintf(":%d", s.config.Port),
+		Handler: &muxHandler{
+			grpcServer:  s.createGRPCServer(),
+			passthrough: s.createHttpServicesHandler(),
 		},
-	)
+	}
+
+	errChan := make(chan error)
+	go func() {
+		defer close(errChan)
+		err := server.ListenAndServe()
+		if err != nil {
+			switch err {
+			case http.ErrServerClosed:
+				log.Info("shutting down server")
+			default:
+				log.Error(err, "could not start an HTTP Server")
+				errChan <- err
+			}
+		}
+	}()
+	log.Info("starting", "interface", "0.0.0.0", "port", s.config.Port)
+
+	select {
+	case <-stop:
+		log.Info("stopping gracefully")
+		return server.Shutdown(context.Background())
+	case err := <-errChan:
+		return err
+	}
+}
+
+func (s *muxServer) NeedLeaderElection() bool {
+	return false
+}
+
+func SetupServer(rt core_runtime.Runtime) error {
+	config := rt.Config().MonitoringAssignmentServer
+
+	if !config.GrpcEnabled && !config.HttpEnabled {
+		return nil
+	}
+
+	rm := rt.ReadOnlyResourceManager()
+
+	var grpcServices []GrpcService
+	var httpServices []HttpService
+
+	addGRPCService := func(svc GrpcService) {
+		if config.GrpcEnabled {
+			grpcServices = append(grpcServices, svc)
+		}
+	}
+
+	addHTTPService := func(svc HttpService) {
+		if config.HttpEnabled {
+			httpServices = append(httpServices, svc)
+		}
+	}
```

Comment:
Is there a scenario when we would like to explicitly enable only http or grpc version of mads?

Reply:
The use case would be only exposing HTTP for the (future) native Prometheus SD, which is all that is necessary.  Is the overhead of a gRPC server minimal? If that's the case, I think we can always run both.

Reply:
I'll remove this for now and just say "always-on" like it is currently